encoding/ccf/encode.go

```diff
@@ -0,0 +1,1751 @@
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, CBORTagTypeDefAndValue,
+		// array, 2 items follow
+		0x82,
+	})
+	if err != nil {
+		return err
+	}
+
+	// element 0: typedef
+	err = e.encodeTypeDefs(types, tids)
+	if err != nil {
+		return err
+	}
+
+	// element 1: type and value
+	return e.encodeInlineTypeAndValue(value, tids)
+}
+
+// encodeTypeAndValue encodes type and value as
+// language=CDDL
+// ccf-type-and-value-message =
+//
+//	; cbor-tag-type-and-value
+//	#6.130(inline-type-and-value)
+func (e *Encoder) encodeTypeAndValue(value cadence.Value, tids ccfTypeIDByCadenceType) error {
+	// Encode tag number
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, CBORTagTypeAndValue,
+	})
+	if err != nil {
+		return err
+	}
+
+	return e.encodeInlineTypeAndValue(value, tids)
+}
+
+// encodeTypeAndValueWithNoTag encodes inline type and value as
+// language=CDDL
+// inline-type-and-value = [
+//
+//	type: inline-type,
+//	value: value,
+//
+// ]
+func (e *Encoder) encodeInlineTypeAndValue(value cadence.Value, tids ccfTypeIDByCadenceType) error {
+	// Encode array head of length 2.
+	err := e.enc.EncodeRawBytes([]byte{
+		// array, 2 items follow
+		0x82,
+	})
+	if err != nil {
+		return err
+	}
+
+	runtimeType := value.Type()
+
+	// element 0: inline-type
+	err = e.encodeInlineType(runtimeType, tids)
+	if err != nil {
+		return err
+	}
+
+	// element 1: value
+	return e.encodeValue(value, runtimeType, tids)
+}
+
+// encodeTypeDefs encodes composite/interface type definitions as
+// language=CDDL
+// composite-typedef = [
+//
+//	+(
+//	  struct-type
+//	  / resource-type
+//	  / contract-type
+//	  / event-type
+//	  / enum-type
+//	  / struct-interface-type
+//	  / resource-interface-type
+//	  / contract-interface-type
+//	  )]
+func (e *Encoder) encodeTypeDefs(types []cadence.Type, tids ccfTypeIDByCadenceType) error {
+	// Encode array head with number of type definitions.
+	err := e.enc.EncodeArrayHead(uint64(len(types)))
+	if err != nil {
+		return err
+	}
+
+	for _, typ := range types {
+		switch x := typ.(type) {
+		case cadence.CompositeType:
+			// Encode struct-type, resource-type, contract-type, event-type, or enum-type.
+			err = e.encodeCompositeType(x, tids)
+			if err != nil {
+				return err
+			}
+
+		case cadence.InterfaceType:
+			// Encode struct-interface-type, resource-interface-type, or contract-interface-type.
+			err = e.encodeInterfaceType(x, tids)
+			if err != nil {
+				return err
+			}
+
+		default:
+			panic(fmt.Errorf("unexpected type %s in type definition", typ.ID()))
+		}
+	}
+
+	return nil
+}
+
+// encodeValue traverses the object graph of the provided value and encodes it as
+// language=CDDL
+// value =
+//
+//	ccf-type-and-value-message
+//	/ simple-value
+//	/ optional-value
+//	/ array-value
+//	/ dict-value
+//	/ composite-value
+//	/ path-value
+//	/ capability-value
+//	/ function-value
+//	/ type-value
+//
```

Comment:
The CCF codec includes code comments that describe parts of [CCF Specification (RC1)](https://github.com/fxamacker/ccf_draft/commit/4e8f42db29925bd15481301729eca1b52852dcb4).  Two notations from RFC 8610 are used:

- [Concise Data Definition Language (CDDL)](https://www.rfc-editor.org/rfc/rfc8610.html) notation is used to document encoding/decoding functions and can help keep the codec and specs in sync.

- [Extended Diagnostic Notation (EDN)](https://www.rfc-editor.org/rfc/rfc8610.html#appendix-G) is used to document parts of unit tests.